### PR TITLE
usbredir_tcp: add usb-redir via tcp testing case

### PR DESCRIPTION
**Pull request by ybduan**
ID: 1911579

Signed-off-by: ybduan <yduan@redhat.com>

### CODE REVIEW

**Comment by ybduan** [comment]
Please help review this PR, thanks! @yanan-fu

**Comment by ybduan** [comment]
(1/2) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.nec-xhci: PASS (53.58 s)
 (2/2) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.qemu-xhci: PASS (50.64 s)

File: qemu/tests/usb_redir.py [hunk 1]
```diff
@@ -72,6 +84,35 @@ def _host_config_check():
...
                 err_msg += "output: %s" % o
                 return (status, err_msg)
+        elif backend == 'tcp_socket':
+            usbredir_cmd = "rpm -q usbredir || yum install -y usbredir"
+            s, o = process.getstatusoutput(usbredir_cmd, shell=True)
+            if s:
+                status = False
+                err_msg = "Fail to install 'usbredir' on host, "
+                err_msg += "output: %s" % o
+                return (status, err_msg)
```
Comment (yanan-fu) [inline qemu/tests/usb_redir.py:94]: Suggest 
```
if not utils_package.package_install("usbredir"):
    ***
    ***
```

File: qemu/tests/cfg/usb.cfg [hunk 2]
```diff
@@ -382,6 +382,10 @@
                 - with_negative_config:
                     display = spice
                     usbredir_unconfigured_value = -1
+                - via_tcp:
+                    chardev_backend_usbredir1 = tcp_socket
+                    chardev_host_usbredir1 = localhost
+                    chardev_port_usbredir1 = 1234
```
Comment (yanan-fu) [inline qemu/tests/cfg/usb.cfg:388]: Suggest get the free port automatically

File: qemu/tests/usb_redir.py [hunk 3]
```diff
@@ -72,6 +84,35 @@ def _host_config_check():
...
+                    err_msg += "output: %s" % o
+                    return (status, err_msg)
+                usbredirserver_cmd = "rpm -i usbredir-server*"
+                s, o = process.getstatusoutput(usbredirserver_cmd, shell=True)
+                if s:
+                    status = False
+                    err_msg = "Fail to install 'usbredir-server' on host, "
+                    err_msg += "output: %s" % o
+                    return (status, err_msg)
+            _start_usbredir_server()
```
Comment (yanan-fu) [inline qemu/tests/usb_redir.py:115]: IMO, it is not belong to `_host_config_check`,  it is a step on the same level as `_start_spice_redirection`. so i suggest:
```
_host_config_check
_start_usbredir_server
_usbredir_preprocess
```

File: qemu/tests/usb_redir.py [hunk 4]
```diff
@@ -72,6 +84,35 @@ def _host_config_check():
...
+                err_msg = "Fail to install 'usbredir' on host, "
+                err_msg += "output: %s" % o
+                return (status, err_msg)
+            usbredirserver_query = "rpm -q usbredir-server"
+            s, o = process.getstatusoutput(usbredirserver_query, shell=True)
+            if s:
+                usbredir_query = "rpm -q usbredir"
+                usbredir_ver = process.getoutput(usbredir_query, shell=True)
+                (build, arch) = usbredir_ver.rsplit(".", 1)
+                brew_cmd = "brew download-build --arch=%s %s" % (arch, build)
```
Comment (yanan-fu) [inline qemu/tests/usb_redir.py:101]: `brew` is internal, it is better do not use it. Have you tried to find a repo that include `usbredir-server` pacakge ?

File: qemu/tests/usb_redir.py (hunk 1 above)
Comment (ybduan) [reply qemu/tests/usb_redir.py:91]: Updated, thanks!

File: qemu/tests/cfg/usb.cfg (hunk 2 above)
Comment (ybduan) [reply qemu/tests/cfg/usb.cfg:388]: Updated, thanks!

File: qemu/tests/usb_redir.py (hunk 3 above)
Comment (ybduan) [reply qemu/tests/usb_redir.py]: Updated, thanks!

**Comment by ybduan** [comment]
(1/2) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.nec-xhci: PASS (53.46 s)
 (2/2) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.qemu-xhci: PASS (51.86 s)

File: qemu/tests/usb_redir.py (hunk 4 above)
Comment (ybduan) [reply qemu/tests/usb_redir.py]: Updated, thanks!

**Comment by ybduan** [comment]
(1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.nec-xhci: PASS (54.83 s)
 (2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.qemu-xhci: PASS (48.70 s)

**Comment by ybduan** [comment]
Please help review this PR, thanks! @yanan-fu

File: qemu/tests/usb_redir.py [hunk 5]
```diff
@@ -258,6 +298,8 @@ def _usb_stick_io(mount_point):
         return
 
     if not _usb_dev_verify():
+        if backend == 'tcp_socket':
+            process.getoutput("killall usbredirserver")
```
Comment (yanan-fu) [inline qemu/tests/usb_redir.py:302]: It is just a clean step, suggest  `process.system(***, ignore_status=True)`

File: qemu/tests/usb_redir.py [hunk 6]
```diff
@@ -270,4 +312,7 @@ def _usb_stick_io(mount_point):
             if iozone_test:
                 iozone_test.clean()
 
+    if backend == 'tcp_socket':
+        process.getoutput("killall usbredirserver")
```
Comment (yanan-fu) [inline qemu/tests/usb_redir.py:316]: Same with above

File: qemu/tests/usb_redir.py (hunk 5 above)
Comment (ybduan) [reply qemu/tests/usb_redir.py]: Updated.

File: qemu/tests/usb_redir.py (hunk 6 above)
Comment (ybduan) [reply qemu/tests/usb_redir.py]: Updated.

**Comment by ybduan** [comment]
(1/2) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.nec-xhci: PASS (63.64 s)
 (2/2) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.qemu-xhci: PASS (50.70 s)

**Comment by ybduan** [comment]
Please help review this PR, thanks! @yanan-fu

**Comment by ybduan** [comment]
(1/2) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.nec-xhci: PASS (52.94 s)
 (2/2) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.usb.usb_redir.via_tcp.without_usb_hub.qemu-xhci: PASS (50.51 s)

**Review by yanan-fu — APPROVED** [review]
Acked-by: Yanan Fu <yfu@redhat.com>

**Pull request merged**